### DISCUSSION TITLE
[B-L072Z-LRWAN1] - RTC alarm failure when timeout <= 2

### DISCUSSION THREAD

**Discussion by jelledevleeschouwer** [Lora-net/LoRaMac-node#1002] [Q&A] [answered]
Dear,

We found ourselves in a situation where an RTC alarm fails randomly if and only if, the alarm is scheduled less or equal than 2 RTC subseconds in the future. To schedule this alarm we directly use `RtcStartAlarm` from rtc-board.c: 
https://github.com/Lora-net/LoRaMac-node/blob/184039f85e0e2c1d075c3881d73b38963c33f5db/src/boards/B-L072Z-LRWAN1/rtc-board.c#L324 Despite the fact we haven't found the root cause for this problem yet, it appears you had a somewhat similar experience. Indeed, on the following lines https://github.com/Lora-net/LoRaMac-node/blob/b0a91e03a25c2d71d46a5b12b7e69ed08360d398/src/system/timer.c#L335 and https://github.com/Lora-net/LoRaMac-node/blob/b0a91e03a25c2d71d46a5b12b7e69ed08360d398/src/system/timer.c#L367 of timer.c, the code invokes a function called `RtcGetMinimumTimeout`. Where `RtcGetMinimumTimeout` defined in rtc-board.c https://github.com/Lora-net/LoRaMac-node/blob/184039f85e0e2c1d075c3881d73b38963c33f5db/src/boards/B-L072Z-LRWAN1/rtc-board.c#L235-L238 returns the magic value of `3`:
https://github.com/Lora-net/LoRaMac-node/blob/184039f85e0e2c1d075c3881d73b38963c33f5db/src/boards/B-L072Z-LRWAN1/rtc-board.c#L39
The return value of the function call is used to prevent any RTC alarm to be configured with an RTC subsecond value any less than 3 in timer.c. The introduction of this `RtcGetMinimumTimeout` we traced back to commit 9091bb2 dating from July 2, 2018 augmented by a comment _Originally posted by @mluis1 in [#473 (comment)](https://github.com/Lora-net/LoRaMac-node/issues/473#issuecomment-399356938)_.

The only plausible explanation that we have for this issue until now, is that ST mentions in the reference manual in
section **26.4.7 RTC initialization and configuration**, in the paragraph about 'Programming the alarm' on page 616 [[RM0376, DocID025941 Rev 5]](https://www.st.com/resource/en/reference_manual/dm00108281-ultralowpower-stm32l0x2-advanced-armbased-32bit-mcus-stmicroelectronics.pdf), the following:

      > Each change of the RTC_CR register is taken into account after
      > around 2 RTCCLK clock cycles due to clock synchronization.

But, according to the block diagram on page 610 (26.4.1) of that same document, RTCCLK clearly points to the LSE input which is 32.768 kHz. So we would expect a synchronization delay of `2 / 32768 s` , which would clearly not have an impact on a RTC alarm scheduled 2 subseconds (`2 / (32768 / (1 << ( 15 - N_PREDIV_S ))) = 2 / 1024 s`) in the future. However, if ST has made a mistake here, and RTCCLK should actually refer to the asynchronously prescaled clock 'ck_apre'. That could perfectly explain why an RTC alarm of 2 RTC subseconds in the future would very randomly fail. 

Do any of you guys know why exactly that `RtcGetMinimumTimeout` was introduced? Because this smells like we have experienced exactly the same issue here.

**Comment by jelledevleeschouwer** [comment, marked as answer]
Hi @mluis1,

Do you perhaps have any feedback on this topic?

Thanks in advance,
Best regards,
Jelle

**Reply by mluis1** [reply]
The feedback that I received is that we cannot setup an alarm with a timeout smaller than 3.

This means that every time the `RtcStartAlarm` API is called one must ensure that `timeout` parameter has at least a value of 3.

I am sorry for not being able to provide you a better answer.

**Reply by jelledevleeschouwer** [reply]
Hi @mluis1,

No problem, thanks for the feedback anyway. We'll discuss this question with ST directly then.

Best regards,
Jelle
